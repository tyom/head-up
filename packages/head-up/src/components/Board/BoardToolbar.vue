<template>
  <div class="BoardToolbar">
    <form class="toolbar-container" @submit.prevent="handleSubmit">
      <div class="toolbar-item _info">
        <input
          v-if="isEditable"
          v-model="boardTitle"
          type="text"
          class="toolbar-input"
          placeholder="Board title"
        >
        <span v-else class="toolbar-title"> {{ title }} (read-only) </span>
      </div>
      <div class="toolbar-item _actions">
        <button v-if="isEditable" :disabled="!hasChanged">Apply</button>
        <button @click.prevent="handleEditDone">OK</button>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      boardTitle: this.title,
    };
  },
  computed: {
    hasChanged() {
      return this.boardTitle !== this.title;
    },
    isEditable() {
      return this.$parent.editable;
    },
  },
  inject: ['handleEditDone'],
  methods: {
    handleSubmit() {
      this.$emit('save', {
        title: this.boardTitle,
      });
    },
  },
};
</script>

<style scoped>
.BoardToolbar {
  background: radial-gradient(
      circle at top left,
      rgba(#20416d, 0.3),
      rgba(#20416d, 0.6)
    )
    no-repeat;
  height: 3.5rem;
  display: flex;
  align-items: stretch;
  position: relative;
}

.toolbar-container {
  flex: 1;
  display: flex;
  align-items: stretch;
}

.toolbar-item {
  display: flex;
  align-items: center;
  padding: 0.6rem 0.5em;
}

.toolbar-title {
  font-size: 1rem;
  padding: 0.5rem;
}

.toolbar-input {
  background: rgba(#fff, 0.1);
  color: #fff;
  border-radius: 0.3rem;
  padding: 0.5rem;
  min-width: 10em;
  width: 100%;
  border: 0;
  font-size: 1rem;
  text-align: inherit;
  outline: 0;
  transition: 0.2s;

  &:hover {
    background: rgba(#fff, 0.15);
  }

  &:focus {
    color: #000;
    background-color: #fff6;
    box-shadow: inset 0 0 2px 1px #0004;
  }

  &::placeholder {
    color: #fff4;
  }

  @media (min-width: 50em) {
    max-width: 50%;
  }
}

.toolbar-item._info {
  flex: 1;
}

.toolbar-item._actions {
  margin-left: auto;

  & button {
    border-radius: 0.3rem;
    padding: 0.6rem 1em;
    font-size: 0.9rem;
    line-height: 1;
    background: #0005;
    border: 1px solid #fff3;
    color: #fff;
    outline: 0;

    &:hover {
      border-color: #fff5;
    }

    &:focus {
      border-color: #fff8;
    }

    &:active {
      background: #000;
    }

    &[disabled] {
      opacity: 0.4;
      cursor: default;
    }
  }

  & button + button {
    margin-left: 0.5em;
  }
}
</style>
