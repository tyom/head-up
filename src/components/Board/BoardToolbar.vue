<template>
  <div class="BoardToolbar">
    <form class="toolbar-container" @submit.prevent="$emit('save', boardState)">
      <div class="toolbar-item _info">
        <input
          v-if="boardState.editable"
          v-model="boardState.title"
          type="text"
          class="toolbar-input"
          placeholder="Board title"
        />
        <span v-else class="toolbar-title">
          {{ boardState.title }} (read-only)
        </span>
      </div>
      <div class="toolbar-item _actions">
        <button v-if="boardState.editable" :disabled="!hasChanged" class="save">
          Apply
        </button>
        <button class="done" @click.prevent="$emit('done', boardState)">
          OK
        </button>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  props: {
    board: {
      type: Object,
      default: () => {},
    },
  },
  data() {
    return {
      boardState: { ...this.board },
    };
  },
  computed: {
    hasChanged() {
      return this.boardState.title !== this.board.title;
    },
  },
};
</script>

<style scoped>
.BoardToolbar {
  background: radial-gradient(
      circle at top left,
      rgba(#20416d, 0.3),
      rgba(#20416d, 0.6)
    )
    no-repeat;
  height: 3.5rem;
  display: flex;
  align-items: stretch;
  position: relative;
}

.toolbar-container {
  flex: 1;
  display: flex;
  align-items: stretch;
}

.toolbar-item {
  display: flex;
  align-items: center;
  padding: 0.6rem 0.5em;
}

.toolbar-title {
  font-size: 1rem;
  padding: 0.5rem;
}

.toolbar-input {
  background: rgba(#fff, 0.1);
  color: #fff;
  border-radius: 0.3rem;
  padding: 0.5rem;
  min-width: 10em;
  width: 100%;
  border: 0;
  font-size: 1rem;
  text-align: inherit;
  outline: 0;
  transition: 0.2s;

  &:hover {
    background: rgba(#fff, 0.15);
  }

  &:focus {
    color: #000;
    background-color: #fff6;
    box-shadow: inset 0 0 1px 1px #0006, 0 0 0 1px #fffa;
  }

  &::placeholder {
    color: #fff4;
  }

  @media (min-width: 50em) {
    max-width: 50%;
  }
}

.toolbar-item._info {
  flex: 1;
}

.toolbar-item._actions {
  margin-left: auto;

  & button {
    border-radius: 0.3rem;
    padding: 0.6rem 1em;
    font-size: 0.9rem;
    line-height: 1;
    background: #0005;
    border: 1px solid #fff3;
    color: #fff;
    outline: 0;

    &:hover {
      border-color: #fff5;
    }

    &:focus {
      border-color: #fff8;
    }

    &:active {
      background: #000;
    }

    &[disabled] {
      opacity: 0.4;
      cursor: default;
    }
  }

  & button + button {
    margin-left: 0.5em;
  }
}
</style>
