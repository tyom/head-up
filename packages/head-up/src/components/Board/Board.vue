<template>
  <div :class="boardClass" class="Board">
    <transition name="slideDown">
      <BoardToolbar
        v-if="!isThumb && isEditing()"
        :board="$props"
      />
    </transition>
    <div :class="layoutClass" class="cells">
      <template v-if="isThumb">
        <div
          v-for="(cell, idx) in cells"
          :key="idx"
          class="cell-placeholder"
        />
      </template>
      <slot v-else-if="!cells.length" />
      <template v-else>
        <Cell
          v-for="(cell, idx) in cells"
          v-bind="cell"
          :editable="editable"
          :key="idx"
          class="cell"
        />
      </template>
    </div>
  </div>
</template>

<script>
import Cell from '../Cell';
import BoardToolbar from './BoardToolbar';

export default {
  name: 'Board',
  components: {
    Cell,
    BoardToolbar,
  },
  props: {
    id: {
      type: String,
      default: '',
    },
    isThumb: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
    cells: {
      type: Array,
      default: () => [],
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boardClass() {
      return {
        _thumb: this.isThumb,
      };
    },
    layoutClass() {
      const cells = this.$slots.default ? this.$slots.default : this.cells;
      if (!cells.length) {
        return;
      }
      return [`u-grid-${cells.length}-x`];
    },
  },
  inject: ['isEditing'],
};
</script>

<style scoped>
.Board {
  display: flex;
  flex-direction: column;
  flex: 1;
  height: 100%;
}

.cell-placeholder {
  background-color: rgba(#fff, 0.1);
}

.cells {
  height: 100%;
  grid-gap: 0.5em;
  padding: 0.5em;
}

._thumb {
  font-size: 0.8em;
  margin: auto;

  & .cells {
    padding: 2px;
    grid-gap: 2px;
  }

  & .cell {
    border: 0;
    background-color: #fff2;
  }
}

.slideDown-enter-active,
.slideDown-leave-active {
  transition: 0.1s;
  overflow: hidden;
}

.slideDown-enter,
.slideDown-leave-to {
  height: 0;
  opacity: 0;
}
</style>
