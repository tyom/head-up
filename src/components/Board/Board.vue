<template>
  <div :id="boardId" :class="boardClass" class="Board">
    <transition name="slideDown">
      <BoardToolbar
        v-if="!isThumb && editMode"
        :board="$props"
        @save="handleSave"
        @done="handleDone"
      />
    </transition>
    <div :class="layoutClass" class="cells">
      <template v-if="isThumb">
        <div v-for="(cell, idx) in cells" :key="idx" class="cell-placeholder" />
      </template>
      <template v-else>
        <slot v-if="slotCells" />
        <Cell
          v-else
          v-for="(cell, idx) in cells"
          v-bind="cell"
          :editable="editable"
          :key="idx"
          class="cell"
        />
      </template>
    </div>
  </div>
</template>

<script>
import get from 'lodash/get';
import store from '../../store';
import Cell from '../Cell';
import BoardToolbar from './BoardToolbar';

export default {
  name: 'Board',
  components: {
    Cell,
    BoardToolbar,
  },
  props: {
    id: {
      type: String,
      required: true,
    },
    isThumb: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
    cells: {
      type: Array,
      default: () => [],
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boardId() {
      return this.isThumb ? null : `board-${this.id}`;
    },
    boardClass() {
      return {
        _thumb: this.isThumb,
      };
    },
    editMode() {
      return store.state.editMode;
    },
    slotCells() {
      if (!this.$slots.default) {
        return;
      }
      const cells = this.$slots.default.filter(
        (x) => get(x, 'componentOptions.tag') === 'Cell',
      );
      return cells.length ? cells : undefined;
    },
    layoutClass() {
      const cells = this.slotCells || this.cells;
      if (!cells.length) {
        return;
      }
      return [`u-grid-${cells.length}-x`];
    },
  },
  methods: {
    handleSave(newState) {
      store.dispatch('UPDATE_BOARD', {
        boardId: this.id,
        payload: newState,
      });
    },
    handleDone(newState) {
      if (this.editable) {
        this.handleSave(newState);
      }
      store.dispatch('TOGGLE_EDIT_MODE', false);
    },
  },
};
</script>

<style scoped>
.Board {
  @apply flex flex-grow flex-col h-full;
}

.cell-placeholder {
  background-color: rgba(#fff, 0.1);
}

.cells {
  height: 100%;
  grid-gap: 0.5em;
  padding: 0.5em;
  overflow: auto;
}

._thumb {
  font-size: 0.8em;
  margin: auto;

  & .cells {
    padding: 2px;
    grid-gap: 2px;
  }

  & .cell {
    border: 0;
    background-color: #fff2;
  }
}

.slideDown-enter-active,
.slideDown-leave-active {
  transition: 0.1s;
  overflow: hidden;
}

.slideDown-enter,
.slideDown-leave-to {
  height: 0;
  opacity: 0;
}
</style>
