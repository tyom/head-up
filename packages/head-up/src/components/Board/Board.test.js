import { shallowMount } from '@vue/test-utils';
import Board from './Board';

const commonSettings = {
  provide: {
    isEditing: () => false,
  },
  stubs: ['Cell', 'BoardToolbar'],
};

test('render empty', () => {
  const wrapper = shallowMount(Board, {
    ...commonSettings,
  });
  expect(wrapper).toMatchSnapshot();
});

test('render normal', () => {
  const wrapper = shallowMount(Board, {
    ...commonSettings,
    propsData: {
      title: 'Board title',
    },
    slots: {
      default: '<Cell/><Cell/>',
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render as thumbnail', () => {
  const wrapper = shallowMount(Board, {
    ...commonSettings,
    propsData: {
      title: 'Board title',
      isThumb: true,
    },
    slots: {
      default: '<Cell/><Cell/>',
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render in edit mode', () => {
  const wrapper = shallowMount(Board, {
    ...commonSettings,
    propsData: {
      title: 'Board title',
      editable: true,
    },
    provide: {
      isEditing: () => true,
    },
    slots: {
      default: '<Cell/><Cell/>',
    },
  });
  expect(wrapper).toMatchSnapshot();
});
