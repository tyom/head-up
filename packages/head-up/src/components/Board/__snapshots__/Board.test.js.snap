// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render as thumbnail 1`] = `
<div class="Board _thumb">
  <!---->
  <div class="cells u-grid-2-x"></div>
</div>
`;

exports[`render empty 1`] = `
<div class="Board">
  <!---->
  <div class="cells"></div>
</div>
`;

exports[`render in edit mode 1`] = `
<div class="Board">
  <boardtoolbar-stub board="[object Object]" name="slideDown"></boardtoolbar-stub>
  <div class="cells u-grid-2-x">
    <cell-stub title="" content=""></cell-stub>
    <cell-stub title="" content=""></cell-stub>
  </div>
</div>
`;

exports[`render normal 1`] = `
<div class="Board">
  <!---->
  <div class="cells u-grid-2-x">
    <cell-stub title="" content=""></cell-stub>
    <cell-stub title="" content=""></cell-stub>
  </div>
</div>
`;
